File: src/ProgenySelfRefImputer.cpp
#include <cmath>
#include "../include/ProgenySelfRefImputer.h"
#include "../include/common.h"

using namespace std;


//////////////////// ProgenySelfRefImputer ////////////////////

ProgenySelfRefImputer::ProgenySelfRefImputer(const vector<GenoRecord *>& rs,
												const Map& map_, double w) :
									VCFHMMSelfRef(map_, w),
									records(rs),
									Cc(calc_Cc(rs)) { }

vector<double> ProgenySelfRefImputer::calc_Cc(
							const vector<GenoRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

double ProgenySelfRefImputer::emission_probability(size_t i, size_t j, int h,
														int parent_gt) const {
	const int	gt = records[i]->get_geno(j+1);
	const int	oc = gt == Genotype::NA ? 4 : gt & 3;
	const int	phased_gt = gt_by_haplotypes(h, parent_gt);
	return E[phased_gt][oc];
}

vector<VCFHMMSelfRef::DP> ProgenySelfRefImputer::initialize_dp(size_t j) const {
	vector<DP>	dp(M(), DP(4, pair<double, int>(MIN_PROB, 0)));
	const GenoRecord	*record = records[0];
	const int	parent_gt = record->get_geno(0);
	for(int h = 0; h < 4; ++h) {	// hidden state
		const double	E = emission_probability(0, j, h, parent_gt);
		dp[0][h] = make_pair(E, h);
	}
	return dp;
}

void ProgenySelfRefImputer::update_dp(size_t i, size_t j,
											vector<DP>& dp) const {
	const GenoRecord	*record = records[i];
	const int	parent_gt = record->get_geno(0) & 3;
	
	for(int h = 0; h < 4; ++h) {
		const double	E = emission_probability(i, j, h, parent_gt);
		
		for(int prev_h = 0; prev_h < 4; ++prev_h) {
			const double	T = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (T + E);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void ProgenySelfRefImputer::update_genotypes(size_t j, const vector<int>& hs) {
	for(size_t i = 0; i < this->M(); ++i) {
		const GenoRecord	*record = records[i];
		const int	parent_gt = record->get_geno(0);
		const int	gtc_int = gt_by_haplotypes(hs[i], parent_gt);
		records[i]->set_geno(j+1, gtc_int | 4);
	}
}

void ProgenySelfRefImputer::impute(size_t j) {
	// DP
	vector<DP>	dp = initialize_dp(j);
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, j, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(j, hs);
}

File: python/ProgenySelfRefImputer.py
from __future__ import annotations

# coding: utf-8
# ProgenyRefImputer.py
# 親がphasedのときに後代をimputeする
# リファレンスがあるとき用
# phasedの遺伝子型を考慮する

from math import log
from typing import List, Tuple

from GenoRecord import GenoRecord
from VCFHMMSelfRef import *
from Genotype import Genotype


#################### ProgenySelfRefImputer ####################

MIN_PROB = -1e300

class ProgenySelfRefImputer(VCFHMMSelfRef[GenoRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[GenoRecord],
									map_: Map, w: float) -> None:
		VCFHMMSelfRef.__init__(self, records, map_)
		self.records: list[GenoRecord] = records
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(records, records[1:]) ]
	
	def num_progenies(self) -> int:
		return len(self.records[0].geno) - 2
	
	# ハプロタイプを決めたときのGenotype
	def gt_by_haplotypes(self, hc: int, parent_gt: int) -> int:
		hc1 = hc & 1
		hc2 = hc >> 1
		return ((parent_gt >> hc1) & 1) | (((parent_gt >> hc2) & 1) << 1)
	
	# i: record index, j: progeny index
	def emission_probability(self, i: int, j: int, h: int,
											parent_gt: int) -> float:
		record = self.records[i]
		gt = record.geno[j+1]
		oc = 4 if gt == Genotype.NA else gt & 3
		phased_gt = self.gt_by_haplotypes(h, parent_gt)
		return self.E[phased_gt][oc]
	
	def progeny_transition_probability(self, i: int,
											prev_hc: int, hc: int) -> float:
		cc = self.Cc[i-1]	# 後代の遷移確率
		hc1 = hc & 1
		hc2 = hc >> 1
		prev_hc1 = prev_hc & 1
		prev_hc2 = prev_hc >> 1
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if prev_hc1 != hc1 else 1.0 - cc) +
				log(cc if prev_hc2 != hc2 else 1.0 - cc))
	
	def initialize_dp(self, j: int, M: int) -> list[DP]:
		dp = [ [ (MIN_PROB, 0) ] * 4 for _ in range(M) ]
		record = self.records[0]
		parent_gt = record.geno[0]
		for h in range(4):		# hidden state
			E = self.emission_probability(0, j, h, parent_gt)
			dp[0][h] = (E, h)
		return dp
	
	def update_dp(self, i: int, j: int, dp: list[DP]) -> None:
		record = self.records[i]
		parent_gt = record.geno[0] & 3
		
		for hc in range(4):		# hidden state
			E_all = self.emission_probability(i, j, hc, parent_gt)
			
			cc = self.Cc[i-1]	# transition probability
			hc1 = hc & 1
			hc2 = hc >> 1
			
			for prev_hc in range(4):
				prev_hc1 = prev_hc & 1
				prev_hc2 = prev_hc >> 1
				# 遷移確率 0なら(1-c)、1ならcを掛ける
				Tc = (log(cc if prev_hc1 != hc1 else 1.0 - cc) +
					  log(cc if prev_hc2 != hc2 else 1.0 - cc))
				
				prob = dp[i-1][prev_hc][0] + (Tc + E_all)
				dp[i][hc] = max(dp[i][hc], (prob, prev_hc))
	
	def update_genotypes(self, j: int, hs: list[int]) -> None:
		M = len(self.records)
		for i in range(M):
			record = self.records[i]
			parent_gt = record.geno[0]
			gtc_int = self.gt_by_haplotypes(hs[i], parent_gt)
			record.geno[j+1] = gtc_int | 4
	
	# j: index of progeny
	def impute(self, j: int) -> None:
		M = len(self.records)	# マーカー数
		
		# DP
		# 外側から、マーカー、ハプロタイプの状態
		dp = self.initialize_dp(j, M)
		for i in range(1, M):
			self.update_dp(i, j, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(j, hs)
